File: crates/minml-core/src/cpu/kernels.rs
// CPU kernels: add, mul, dot, ones, gather. Direct port of cpu_backend.cpp.
use crate::array::Array;
use crate::cpu::backend::{with_f32, with_f32_mut, with_i32, with_i32_mut};
use crate::dtype::DType;
use crate::error::{MinmlError, Result};

pub fn add(a: &Array, b: &Array, out: &Array) -> Result<()> {
    let n = out.size();
    let buf_a = a.buffer().expect("evaluated");
    let buf_b = b.buffer().expect("evaluated");
    let buf_out = out.buffer().expect("evaluated");
    with_f32(&*buf_a, |pa| {
        with_f32(&*buf_b, |pb| {
            with_f32_mut(&*buf_out, |po| {
                for i in 0..n {
                    po[i] = pa[i] + pb[i];
                }
            })
        })
    });
    Ok(())
}

pub fn mul(a: &Array, b: &Array, out: &Array) -> Result<()> {
    let n = out.size();
    let buf_a = a.buffer().expect("evaluated");
    let buf_b = b.buffer().expect("evaluated");
    let buf_out = out.buffer().expect("evaluated");
    with_f32(&*buf_a, |pa| {
        with_f32(&*buf_b, |pb| {
            with_f32_mut(&*buf_out, |po| {
                for i in 0..n {
                    po[i] = pa[i] * pb[i];
                }
            })
        })
    });
    Ok(())
}

pub fn dot(a: &Array, b: &Array, out: &Array) -> Result<()> {
    let n = a.size();
    let buf_a = a.buffer().expect("evaluated");
    let buf_b = b.buffer().expect("evaluated");
    let buf_out = out.buffer().expect("evaluated");
    let mut sum = 0.0f32;
    with_f32(&*buf_a, |pa| {
        with_f32(&*buf_b, |pb| {
            for i in 0..n {
                sum += pa[i] * pb[i];
            }
        })
    });
    with_f32_mut(&*buf_out, |po| po[0] = sum);
    Ok(())
}

pub fn ones(out: &Array) -> Result<()> {
    let n = out.size();
    let buf_out = out.buffer().expect("evaluated");
    match out.dtype() {
        DType::Float32 => with_f32_mut(&*buf_out, |p| {
            for v in &mut p[..n] {
                *v = 1.0;
            }
        }),
        DType::Int32 => with_i32_mut(&*buf_out, |p| {
            for v in &mut p[..n] {
                *v = 1;
            }
        }),
    }
    Ok(())
}

pub fn gather(table: &Array, indices: &Array, out: &Array) -> Result<()> {
    let big_n = table.shape()[0];
    let mut trail: usize = 1;
    for d in &table.shape()[1..] {
        trail *= *d;
    }
    let m = indices.size();
    let buf_t = table.buffer().expect("evaluated");
    let buf_idx = indices.buffer().expect("evaluated");
    let buf_out = out.buffer().expect("evaluated");

    let mut err: Option<MinmlError> = None;
    with_f32(&*buf_t, |t| {
        with_i32(&*buf_idx, |idx| {
            with_f32_mut(&*buf_out, |o| {
                for i in 0..m {
                    let k = idx[i];
                    if k < 0 || (k as usize) >= big_n {
                        err = Some(MinmlError::GatherOob);
                        return;
                    }
                    let src = (k as usize) * trail;
                    let dst = i * trail;
                    o[dst..dst + trail].copy_from_slice(&t[src..src + trail]);
                }
            })
        })
    });
    if let Some(e) = err {
        return Err(e);
    }
    Ok(())
}

File: src/cpu/kernels.ts
import type { Array } from "../array.js";
import { DType } from "../dtype.js";
import { MinmlError } from "../error.js";
import { f32_view, i32_view } from "./backend.js";

function must_buffer(a: Array) {
  const b = a.buffer();
  if (b === null) throw MinmlError.other("kernel: input not evaluated");
  return b;
}

export function add(a: Array, b: Array, out: Array): void {
  const n = out.size();
  const pa = f32_view(must_buffer(a));
  const pb = f32_view(must_buffer(b));
  const po = f32_view(must_buffer(out));
  for (let i = 0; i < n; i++) {
    po[i] = pa[i] + pb[i];
  }
}

export function mul(a: Array, b: Array, out: Array): void {
  const n = out.size();
  const pa = f32_view(must_buffer(a));
  const pb = f32_view(must_buffer(b));
  const po = f32_view(must_buffer(out));
  for (let i = 0; i < n; i++) {
    po[i] = pa[i] * pb[i];
  }
}

export function dot(a: Array, b: Array, out: Array): void {
  const n = a.size();
  const pa = f32_view(must_buffer(a));
  const pb = f32_view(must_buffer(b));
  const po = f32_view(must_buffer(out));
  let sum = 0.0;
  for (let i = 0; i < n; i++) {
    sum += pa[i] * pb[i];
  }
  po[0] = sum;
}

export function ones(out: Array): void {
  const n = out.size();
  const buf = must_buffer(out);
  switch (out.dtype()) {
    case DType.Float32: {
      const p = f32_view(buf);
      for (let i = 0; i < n; i++) p[i] = 1.0;
      break;
    }
    case DType.Int32: {
      const p = i32_view(buf);
      for (let i = 0; i < n; i++) p[i] = 1;
      break;
    }
  }
}

export function gather(table: Array, indices: Array, out: Array): void {
  const big_n = table.shape()[0];
  let trail = 1;
  for (let d = 1; d < table.shape().length; d++) {
    trail *= table.shape()[d];
  }
  const m = indices.size();
  const t = f32_view(must_buffer(table));
  const idx = i32_view(must_buffer(indices));
  const o = f32_view(must_buffer(out));

  for (let i = 0; i < m; i++) {
    const k = idx[i];
    if (k < 0 || k >= big_n) {
      throw MinmlError.gather_oob();
    }
    const src = k * trail;
    const dst = i * trail;
    for (let j = 0; j < trail; j++) {
      o[dst + j] = t[src + j];
    }
  }
}
